zkml/src/layers/transformer/softmax.rs

```diff
@@ -1,37 +1,192 @@
 //! This layer applies the softmax function to the last dimension of the input tensor
+use std::marker::PhantomData;
+
 use crate::{
-    Element, Tensor,
-    layers::provable::{Evaluate, LayerOut, OpInfo, QuantizeOp},
+    Element, ScalingStrategy, Tensor,
+    layers::provable::{
+        Evaluate, LayerOut, NodeId, OpInfo, ProvingData, QuantizeOp, QuantizeOutput,
+    },
+    quantization::ScalingFactor,
     tensor::Number,
 };
 
-#[derive(Debug, Clone)]
+use anyhow::{Result, anyhow, ensure};
+
+use ff_ext::ExtensionField;
+
+use multilinear_extensions::util::ceil_log2;
+use serde::{Deserialize, Serialize};
+
+/// The base 2 logarithm of the scale factor used in exponential lookup tables
+const LOG_SCALE_FACTOR: usize = 24;
+/// The scale factor for our fixed point arithmetic
+const SCALE_FACTOR: usize = 1 << LOG_SCALE_FACTOR;
+/// The scale factor of the outputs of the `exp` lookup
+const OUTPUT_SCALE_FACTOR: usize = 1 << (LOG_SCALE_FACTOR - 1);
```

Comment:
Can we find another name for these constants? Afaiu these are the inverse if the scaling factors employed usually, so it could be a bit misleading to name these ones as scaling factors,